biz.aQute.bndlib/src/aQute/bnd/plugin/jpms/JPMSModuleInfoPlugin.java

```diff
@@ -0,0 +1,397 @@
+		mainClass(analyzer, writer);
+
+		// TODO use annotations to store other header info???
+		// AnnotationVisitor visitAnnotation = classWriter.visitAnnotation(...,
+		// false);
+
+		ByteBufferDataOutput bbout = new ByteBufferDataOutput();
+		writer.write(bbout);
+
+		analyzer.getJar()
+			.putResource(MODULE_INFO_CLASS,
+				new EmbeddedResource(bbout.toByteArray(), analyzer.lastModified()));
+	}
+
+	private String getModuleName(Analyzer analyzer, Jar jar) throws Exception {
+		String moduleName = jar.getModuleName();
+		if (moduleName == null) {
+			if (jar.getSource() != null && jar.getSource()
+				.isDirectory()) {
+				return null;
+			}
+			moduleName = jar.getName();
+			Matcher matcher = mangledModuleName.matcher(moduleName);
+			if (matcher.matches()) {
+				moduleName = matcher.group(1);
+			}
+			if (logger.isWarnEnabled())
+				logger.warn("Using generated module name '{}' for: {}", moduleName, jar);
+		}
+		return moduleName;
+	}
+
+	private int access(Parameters requireCapabilities) {
+		return requireCapabilities.entrySet()
+			.stream()
+			.filter(entry -> Header.removeDuplicateMarker(entry.getKey())
+				.equals(EXTENDER_NAMESPACE))
+			.findAny()
+			.map(entry -> ModuleAttribute.ACC_OPEN)
+			.orElse(0);
+	}
+
+	private String name(Analyzer analyzer) {
+		return analyzer.getProperty(AUTOMATIC_MODULE_NAME, analyzer.getBsn());
+	}
+
+	private void exportPackages(Analyzer analyzer, ModuleInfoWriter writer) {
+		Packages contained = analyzer.getContained();
+
+		analyzer.getExports()
+			.forEach((packageRef, attrs) -> {
+				String[] targets = EMPTY;
+
+				Attrs containedAttrs = contained.get(packageRef);
+				if (containedAttrs != null && containedAttrs.containsKey(INTERNAL_EXPORT_TO_MODULES_DIRECTIVE)) {
+					targets = Strings.split(containedAttrs.get(INTERNAL_EXPORT_TO_MODULES_DIRECTIVE))
+						.toArray(new String[0]);
+				}
+
+				// TODO Do we want to handle access? I can't think of a reason.
+				// Allowed: 0 | ACC_SYNTHETIC | ACC_MANDATED
+
+				writer.exports(packageRef.getBinary(), 0, targets);
+			});
+	}
+
+	private void mainClass(Analyzer analyzer, ModuleInfoWriter writer) {
+		String mainClass = analyzer.getProperty(MAIN_CLASS);
+
+		if (mainClass != null) {
+			TypeRef typeRef = analyzer.getTypeRefFromFQN(mainClass);
+
+			writer.mainClass(typeRef.getBinary());
+		}
+	}
+
+	private ModuleInfoWriter nameAccessAndVersion(Entry<String, Attrs> instruction, Parameters requireCapability,
+		Analyzer analyzer) {
+		Attrs attrs = instruction.getValue();
+
+		String name = instruction.getKey();
+		// Allowed: 0 | ACC_OPEN | ACC_SYNTHETIC | ACC_MANDATED
+		String access = attrs.computeIfAbsent("access", k -> String.valueOf(access(requireCapability)));
+		String version = attrs.computeIfAbsent("version", k -> analyzer.getVersion());
+
+		return new ModuleInfoWriter(Clazz.JAVA.OpenJDK9.getMajor(), name, version,
+			Integer.valueOf(access) == ModuleAttribute.ACC_OPEN);
```

Comment:
`Integer.parseInt(access) == ModuleAttribute.ACC_OPEN` otherwise you create an `Integer` and then `==` it to an `int` boxed to an `Integer`. Too many `Integer`s which are not necessary.